### ISSUE TITLE
Better handling of missing JSON files

### ISSUE THREAD

**Issue by akirk** [Automattic/wp-cldr#72]
In light of potentially downloading the CLDR files separately, missing files should be handled better. 

Currently the WP CLDR demo page looks like this:
<img width="1041" alt="missing-json" src="https://cloud.githubusercontent.com/assets/203408/13501456/0fb5c3a6-e167-11e5-99c0-72638bebc551.png">

Probably it would be useful to display a notice in WP-Admin about the missing files.

**Comment by akirk** [comment]
Related #37

**Comment by akirk** [comment]
From a coding standpoint I'd suggest to return `true` or `false` in `initialize_locale_bucket()` and introduce a `$this->initalized` class variable and check for that before accessing the `localized` array.

**Comment by akirk** [comment]
As a continuation, you could introduce a `class CLDR_Exception extends Exception {}` and use `throw new CLDR_Exception( 'missing JSON file' )` in the initializing.

**Comment by stuwest** [comment]
In general this makes sense.  I see a few separate issues:
1. Improved handling of the English fallback. Currently, the class tries to always return something for a given locale, even if it falls back to English. There isn't any visibility of the fallback outside of `initialize_locale_bucket()`. The result is that a locale may be "initialized" and in memory, but it could just be with English strings. It would be better to avoid initializing locales with just English data. I'll do that and move the fall back logic into `get_locale_bucket()`.
2. Messaging around missing locale files. Yes it would be good to say check for the currently installed languages (`get_available_languages()`?) and then display a message in plugin settings if any don't have cldr files installed.
3. better exception handling of missing files so we don't have that mess of notices and warnings.

@akirk, how did you get to the screenshot above?  did you have to delete one of the CLDR files or did you get that with a fresh install of the plugin?

**Comment by akirk** [comment]
I renamed the `json` directory. I know it's not quite the fair thing to do but it was intended as preparation for the separate JSON downloading.

**Comment by stuwest** [comment]
OK @akirk I just added another commit https://github.com/Automattic/wp-cldr/commit/102ee31fdc39c8ffa49b02d3dfde256ab529b12a to PR #76. I think it address your issues. Diff is a bit messy but it:

- refactors `get_cldr_json_file()` to create a few new methods -- `get_cldr_json_path()` and `is_cldr_json_available()` -- so that we can access them directly
- refactors `initialize_locale_bucket()` to create a new method -- `get_best_available_cldr_json_locale()` -- that isolates the logic around language variation fallbacks
- shift to public static methods for the above
- add some cleanup `isset()` checks to a few of the newish methods in the class
- fix plugin settings page to use the new methods above and display a message when a JSON file isn't available

something is off with caching so I've disabled it for now I'll figure that out tomorrow.
